lib/cocoa/src/protocol/TCompactProtocol.m

```diff
@@ -0,0 +1,706 @@
+{
+    uint8_t protocolId = [self readByte];
+    if (protocolId != COMPACT_PROTOCOL_ID)
+    {
+        @throw [TProtocolException exceptionWithName:@"TProtocolException" reason:[NSString stringWithFormat:@"Expected protocol id %X but got %X", COMPACT_PROTOCOL_ID, protocolId]];
+    }
+    
+    uint8_t versionAndType = [self readByte];
+    uint8_t version = versionAndType & COMPACT_VERSION_MASK;
+    if (version != COMPACT_VERSION)
+    {
+        @throw [TProtocolException exceptionWithName:@"TProtocolException" reason:[NSString stringWithFormat:@"Expected version %d but got %d", COMPACT_VERSION, version]];
+    }
+    
+    int type = (versionAndType >> COMPACT_TYPE_SHIFT_AMOUNT) & COMPACT_TYPE_BITS;
+    int sequenceID = (int)[self readVarint32];
+    NSString* name = [self readString];
+    
+    if (ptype != NULL)
+    {
+        *ptype = type;
+    }
+    if (psequenceID != NULL)
+    {
+        *psequenceID = sequenceID;
+    }
+    if (pname != NULL)
+    {
+        *pname = name;
+    }
+}
+
+-(void)readStructBeginReturningName:(NSString**)pname
+{
+    [lastField addObject:[NSNumber numberWithShort:lastFieldId]];
+    lastFieldId = 0;
+    
+    if (pname != NULL)
+    {
+        *pname = @"";
+    }
+}
+
+-(void)readStructEnd
+{
+    lastFieldId = [[lastField lastObject] shortValue];
+    [lastField removeLastObject];
+}
+
+-(void)readFieldBeginReturningName:(NSString**)pname type:(int*)pfieldType fieldID:(int*)pfieldID
+{
+    uint8_t byte = [self readByte];
+    uint8_t type = byte & 0x0f;
+    
+    // if it's a stop, then we can return immediately, as the struct is over.
+    if (type == TCType_STOP)
+    {
+        if (pname != NULL)
+        {
+            *pname = @"";
+        }
+        if (pfieldType != NULL)
+        {
+            *pfieldType = TType_STOP;
+        }
+        if (pfieldID != NULL)
+        {
+            *pfieldID = 0;
+        }
+        return;
+    }
+    
+    short fieldId = 0;
+    
+    // mask off the 4 MSB of the type header. it could contain a field id delta.
+    short modifier = (byte & 0xf0) >> 4;
+    if (modifier == 0)
+    {
+        // not a delta. look ahead for the zigzag varint field id.
+        fieldId = [self readI16];
+    }
+    else
+    {
+        // has a delta. add the delta to the last Read field id.
+        fieldId = lastFieldId + modifier;
+    }
+    
+    int fieldType = [self getTTypeForCompactType:type];
+    
+    if (pname != NULL)
+    {
+        *pname = @"";
+    }
+    if (pfieldType != NULL)
+    {
+        *pfieldType = fieldType;
+    }
+    if (pfieldID != NULL)
+    {
+        *pfieldID = fieldId;
+    }
+    
+    // if this happens to be a boolean field, the value is encoded in the type
+    if (type == TCType_BOOLEAN_TRUE || type == TCType_BOOLEAN_FALSE)
+    {
+        // save the boolean value in a special instance variable.
+        booleanValue = [[NSNumber numberWithBool:type == TCType_BOOLEAN_TRUE ? YES : NO] retain_stub];
```

Comment:
The `bool ? YES : NO` bit is superfluous - the result of the test is already a bool. It reduces to `YES ? YES : NO` or `NO ? YES : NO` which reduces to either `YES` or `NO` respectively.